qemu/tests/vioinput_keyboard.py

```diff
@@ -0,0 +1,136 @@
+import json
+import os
+import time
+import logging
+
+from virttest import error_context
+from virttest import utils_test
+from virttest import graphical_console
+from provider import input_event_proxy
+from virttest import data_dir
+
+
+@error_context.context_aware
+def keyboard_test(test, params, vm, key_name='', hold_time=0, interval=0):
+    """
+    Keyboard test. Support single key and combination key tests.
+
+    :param test: kvm test object
+    :param params: dictionary with the test parameters
+    :param vm: vm object
+    :param key_name: tested keyname, default key_name='', will covered all keys
+                     on the keyboard(us).
+    :param hold_time: the time that keep one key pressed, default hole_time=0.
+    :param interval: the interval between two keys, default interval=0.
+    """
+
+    def get_keycode_cfg(filename):
+        """
+        Get keyname to keycode cfg table.
+        :param filename: filename that key to keycode file.
+        """
+        keycode_cfg_path = os.path.join(data_dir.get_deps_dir("key_keycode"), filename)
+        with open(keycode_cfg_path) as f:
+            return json.load(f)
+
+    def key_check(key):
+        """
+        Check received key event match exepected key event.
+        :param key: tested key name.
+        """
+        events_queue = listener.events
+
+        if '-' in key:
+            key_lst = [key_check_cfg[k] for k in key.split('-')]
+        else:
+            key_lst = [key_check_cfg[key]]
+        key_num = len(key_lst)
+        key_event_lst = list()
+
+        while not events_queue.empty():
+            events = events_queue.get()
+            key_event_lst.append((events["keyCode"], events["type"]))
+
+        if len(key_event_lst) < 2 * key_num:
```

Comment:
Doesn't it only work for `hold_time == 0` as well?

Reply:
yes, it work for hold_time=0 or hold_time>0, any way, key_event_list have 2*key_num values at least.